Enumerable methods TSE Review.

carlosveigadev opened the pull request:
The following methods were created and tested.
```
#my_each;
#my_each_with_index;
#my_select;
#my_all?;
#my_any;
#my_none;
#my_count;
#my_map;
#my_inject;
```

I did this project alone because my partner stopped answering due to connectivity issues, so I was requested by Willow to do it by myself.

carlosveigadev committed:
Add main files and linters setup

carlosveigadev committed:
Add my_each method

carlosveigadev committed:
add tests to my_each

carlosveigadev committed:
add my_each_with_index method

carlosveigadev committed:
add my_each_with_index demo

carlosveigadev committed:
add my_select method

carlosveigadev committed:
add my_select demo

carlosveigadev committed:
add my_all and its related methods

carlosveigadev committed:
add my_all demo tests

carlosveigadev committed:
add my_any and its extra methods

carlosveigadev committed:
add my_any tests and demo

carlosveigadev committed:
add my_none method

carlosveigadev committed:
add my_none method tests and demo

carlosveigadev committed:
add my_count method, tests and demo

carlosveigadev committed:
add my_map and demo/tests

carlosveigadev committed:
added my_inject method and tests

carlosveigadev committed:
fixed my_all method not returning right value on one test

carlosveigadev committed:
add fix to some rubocop errors, and rubocop.yml file

carlosveigadev committed:
add module length to rubocop.yml

carlosveigadev committed:
add module length to rubocop.yml

carlosveigadev committed:
exclude test.rb of rubocop

carlosveigadev committed:
fix name of rubocop.yml file

carlosveigadev committed:
fix rubocop issues

carlosveigadev committed:
add configuration to cops to rubocop.yml file

add empty line after guard clause in some lines on main.rb

carlosveigadev committed:
fix rubocop config in .yml

carlosveigadev committed:
fix trailspace error

carlosveigadev committed:
add trail space line

carlosveigadev committed:
disable caseEquality

carlosveigadev committed:
fix rubocop offenses line after guardClause

carlosveigadev committed:
add readme instructions and details

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    array
+  end
+
+  def my_all?(arg = nil)
+    puts "#{__FILE__}:#{__LINE__}: w: block not used" if arg && block_given?
+    return my_all_extra_func(arg) if arg
+
+    if !block_given?
+      size.times { |n| return false if to_a[n] == false || to_a[n].nil? }
+    else
+      my_each { |n| return false unless yield n }
+    end
+    true
+  end
```
# my_all? 
1. [ ] When a Regex is passed as an argument returns false if any of the collection does not match the Regex
### Test 👉  `words.my_all?(/o/)`

|Expected|Got|
|:---:|:---:|
|false|true|

#
2. [ ] When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern.
### Test 👉  `array.my_all?(3)`

|Expected|Got|
|:---:|:---:|
|true|`TypeError`|
#

3. [ ] When a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern.
### Test 👉  `array.my_all?(3)`

|Expected|Got|
|:---:|:---:|
|true|`TypeError`|

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    true
+  end
+
+  def my_any?(arg = nil)
+    if arg
+      puts "#{__FILE__}:#{__LINE__}: w: block not used" if block_given?
+      return my_any_extra_func(arg)
+    elsif block_given?
+      my_each { |n| return true if yield n }
+    else
+      my_each { |n| return true if n }
+    end
+    false
+  end
```
# my_any?
1. [ ] When a class is passed as an argument returns true if at least one of the collection is a member of such class:: Numeric

### Test 👉  `array.my_any?(Numeric)`

|Expected|Got|
|:---:|:---:|
|true|`#<Array:2080> => [2, 4, 7, 3, 7, 1, 3, 5, 7, 7, 4, 4, 2, 7, 3, 3, 4, 2, 5, 2, 1, 3, 4, 5, 8, 2, 6, 0, 5, 0, 0, 4, 5, 1..., 5, 8, 4, 5, 2, 4, 8, 2, 3, 3, 4, 5, 6, 0, 4, 5, 7, 3, 3, 2, 6, 0, 3, 1, 8, 1, 1, 2, 1, 8, 8, 6, 0]`|

#

2. [ ] When a Regex is passed as an argument returns true if any of the collection matches the Regex

### Test 👉 ` words.my_any?(/d/)`

|Expected|Got|
|:---:|:---:|
|true|`#<Array:2100> => ["dog", "door", "rod", "blade"]`|

#
3. [ ]  When a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
### Test 👉  `words.my_any?('cat')`

|Expected|Got|
|:---:|:---:|
|true|false|

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    end
+  end
+
+  def my_none?(arg = nil)
+    if arg
+      puts "#{__FILE__}:#{__LINE__}: w: block not used" if block_given?
+      !my_any?(arg)
+    elsif block_given?
+      !my_any?(&Proc.new)
+    else
+      !my_any?
+    end
+  end
```
# my_none
1. [ ] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

### Test 👉 `words.my_none?(5)`

|Expected|Got|
|:---:|:---:|
|false|true|

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    end
+  end
+
+  def my_count(arg = 0)
+    counter = 0
+    if block_given?
+      my_each { |n| counter += 1 if yield n }
+    elsif arg.zero?
+      my_each { counter += 1 }
+    else
+      my_each { |n| counter += 1 if arg == n }
+    end
+    counter
+  end
```
# my_count
1. [ ] Counts the number of items in enum that are equal to the item if an argument is given
### Test 👉 `array.my_count(LOWEST_VALUE)`

|Expected|Got|
|:---:|:---:|
|`9`|`100`|

christianotieno commented on main.rb:
```diff
@@ -0,0 +1,133 @@
+    mapped
+  end
+
+  def my_inject(prm1 = nil, prm2 = nil)
+    arr = is_a?(Array) ? self : to_a
+    sym = prm1 if prm1.is_a?(Symbol) || prm1.is_a?(String)
+    accum = prm1 if prm1.is_a? Integer
+
+    if prm1.is_a?(Integer)
+      if prm2.is_a?(Symbol) || prm2.is_a?(String)
+        sym = prm2
+      elsif !block_given?
+        raise "#{prm2} is not a symbol nor a string"
+      end
+    elsif prm1.is_a?(Symbol) || prm1.is_a?(String)
+      raise "#{prm2} is not a symbol nor a string" if !prm2.is_a?(Symbol) && !prm2.nil?
+
+      raise "undefined method `#{prm2}' for :#{prm2}:Symbol" if prm2.is_a?(Symbol) && !prm2.nil?
+    end
+
+    if sym
+      arr.my_each { |n| accum = accum ? accum.send(sym, n) : n }
+    elsif block_given?
+      arr.my_each { |n| accum = accum ? yield(accum, n) : n }
+    else
+      raise 'no block given'
+    end
+    accum
+  end
+end
```
# my_inject
1. [ ] Raises a `"LocalJumpError"` when no block or argument is given 
### Test 👉 ` array.my_inject`

|Expected|Got|
|:---:|:---:|
|`LocalJumpError`|`#<RuntimeError: no block given> with backtrace:`|

christianotieno reviewed (changes requested):
# **CHANGES REQUIRED** :warning:
Hey @wrakc   :wave:
-
You have done a splendid job! :tada: Keep it up.
We, however, do have some few changes to carry out before we proceed (see review comment).
##
## NOTE: These are the variables I used in the tests below: 👇 
```
  let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }
  let(:words) { %w[dog door rod blade] }
  let(:range) { Range.new(5, 50) }
  let(:hash) { { a: 1, b: 2, c: 3, d: 4, e: 5 } }
  let(:numbers) { [1, 2i, 3.14] }
  let!(:array_clone) { array.clone }
```
- If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email. :eyes:
##
- When you have made the specified changes, please submit another code review. Thank you!
Happy coding :smile:
**[Christian Otieno](https://github.com/ChristianOtieno)**
[Twitter](https://twitter.com/iamchrisotieno) [linkedin](https://www.linkedin.com/in/christianotieno/) [GitLab](https://gitlab.com/christianotieno)
---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

carlosveigadev committed:
fix my_all not correctly evaluating regex expressions

carlosveigadev committed:
fix my_all? not returning right value when an arg was passed

carlosveigadev committed:
fix my_all? not returning right boolean value

carlosveigadev committed:
fix my_none? returning wrong boolean value

carlosveigadev committed:
fix my_inject return LocalJumpError when no block is given

carlosveigadev committed:
fix rubocop offenses

carlosveigadev commented:
Thanks for the support @christianotieno!

I improved the code as requested:
- Fixed my_all? not correctly returning a boolean value when a regex was passed as an argument;
- Fixed my_all? TypeError when an arg was passed;
- Fixed my_any? returning self when class was passed as an argument or regex;
- Fixed my_any? not returning a right boolean value when a pattern was passed as an argument;
- Fixed my_none? not returning a right boolean value when a pattern was passed as an argument;
- Fixed my_count to count the similar values when a number was passed as an argument;
- Fixed my_inject to return LocalJumpError when a block was not given to the method.

addod19 commented:
## Pull Request under review by Daniel Larbi Addo (TSE)

addod19 commented:
## Pull Request under review by Daniel Larbi Addo (TSE)

addod19 reviewed (changes requested):
## Status: Require changes :recycle: :traffic_light: 

Hi @wrakc, great job with this project :+1: However, there are a few changes we need to implement in order to approve the project. Kindly check comments below: :smile: 
- ```test.rb``` kindly change the ```require``` to ```require_relative```. Since ```require_relative``` is used for files in our current directory
- I test the methods, and this is the result as we compare to the built-in enumerable methods. The screenshot below will help us address these test cases
- There is warning with ```my_none``` method but returns ```true``` as in the ```none``` built-in enumerable method
![NONE](https://user-images.githubusercontent.com/46388113/87194495-d2058600-c2e8-11ea-9e3a-88e4f81ef4ae.png)
![NONE1](https://user-images.githubusercontent.com/46388113/87194963-fc574380-c2e8-11ea-9397-f064f0b69f79.png)


Kindly fix these and submit another code review :+1: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo)

carlosveigadev committed:
fix readme, remove author 2

carlosveigadev committed:
change test.rb require to requie_relative

carlosveigadev committed:
fix my_any and my_none giving boolean error when regex passed as arg

carlosveigadev commented:
Thanks for the support @addod19.

I improved the code as requested:
- Change require to require_relative in the test.rb;
- Fixed the warning about Proc.new deprecated.

rOluochKe reviewed (approved):
Hello Team!
### You got it right this time :clap: the project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next project.
- Happy coding!
